### PR TITLE
Handle needs_authentication response for subscription renewal payment

### PR DESCRIPTION
**Pull request by v18**
# Changes in this PR

When handling subscriptions renewals, this PR handles the case when the response from Stripe is `authentication_required`.

These are all of additions:
- Add order note when an authentication is required by the user
- Saving the payment intent to the order
- Setting the transaction ID to the order
- Mark order status `on-hold`
- Use a new hook called `wc_gateway_stripe_process_payment_authentication_required` so that devs can customize emails, etc

These are removals (what used to happen for an error response, and will not for the `authentication_required` error):
- Logging an error using `WC_Stripe_Logger`
- The hook `wc_gateway_stripe_process_payment_error`
- Setting the order status to `failed`

# Things not handled in this PR
Any customer-related side effects needed for the authentication_required respones:
- Making sure the order will eventually be set to 'failed' if the customer does not come back in time (no PR yet for this)
- Sending the customer an email about the authentication required (this is handled in PR #986  )

# Questions
- [ ] Do we need to reduce stock like we do for `pending` in process_response()?

# Before merging
- [ ] Clean up the commit history
- [x] Test this with PR #986 to make sure everything makes sense together. For example, `process_response()` should work as expected once a payment authentication is received from the customer.
- [x] This should not affect the WC.com, but make sure this assumption is correct before merging. If it does affect WC.com - make sure things work.

_Note: Some notes were here before while the PR was in progress. I'm moving them to a comment for reference._

cc @LCmry @dwainm @DanReyLop @RadoslavGeorgiev

### CODE REVIEW

**Comment by v18** [comment]
What follows is the PR description from earlier while it was still in progress, including possibly useful notes:

---

This is a super preliminary PR, of just my work-in-progress for solving #957. Basically, a jumping off place before I head out on afk.

The changes pushed so far:
1. There is a good API response in json format we can use in a test
2. There is _a_ test, I basically copied it from the success case. Definitely needs to be re-read and amended.
3. There is some logic for figuring out if a response is `needs_authentication`, but not a good amount of work handling it.

Below are some notes I have on `needs_authentication`:

# What should happen for the order when a card requiers authentication?

First, let's see what we do in the other two cases:

### failed order: in the case that an order fails we do this:

- ```$renewal_order->add_order_note($localized_error_message)```
- ```WC_Stripe_Logger::log( 'Error: ' . $e->getMessage() );```
- ```do_action( 'wc_gateway_stripe_process_payment_error', $e, $renewal_order );```
- ```$renewal_order->update_status( 'failed' );```

### successful order: what happens when we process payment?

- ```do_action( 'wc_gateway_stripe_process_payment', $response, $renewal_order );```
- ```$this->process_response( end( $response->charges->data ), $renewal_order );```
  - 🤔 ⬇️ Should we do the follow after a successful authentication? 
  - log: ```WC_Stripe_Logger::log( 'Processing response: ' . print_r( $response, true ) );```
  - ❓store charge data: ```$order->update_meta_data( '_stripe_charge_captured', $captured )``` (but the charge was not captured in our case...)
  - [probably should not be done in needs_authentication] update balance transaction
  - [probably should not be done in needs_authentication] set transaction id: ```$order->set_transaction_id( $response->id )```
  - [probably should not be done in needs_authentication] update stock
  - add order note
  - update status

### ? what should we do for each thing for a 'authentication_required' response?

- do we want to write a new order note - probably yes
- do we want to log it as an error? probably not...
- do we want a new action? probably yes, seems people customize a few things like emails in there ```do_action( 'wc_gateway_stripe_process_payment_authentication_required', $e, $renewal_order );```
- what status do we want? we want `on-hold`: ```$renewal_order->update_status( 'on-hold' );```
- what about saving the payment intent to the order?

### ? what should we do when a user comes back and authenticates after all?

That should happen in a new PR, but... here are some thoughts:

- this happens for a successful order:
  - ```$this->process_response( end( $response->charges->data ), $renewal_order );```
  - ```do_action( 'wc_gateway_stripe_process_payment', $response, $renewal_order );``` 

### things to pay attention to / other considerations 

- one order can have multiple payment attempts
- If a recurring subscription is renewed multiple times, and the last order failed, it'll try to pay on that same order (WC Subs will send the last failed order number to us)
- make sure internal notes are not sent to customer? our notes should be purple or grey

File: includes/compat/class-wc-stripe-subs-compat.php [hunk 1]
```diff
@@ -257,10 +261,31 @@ public function process_subscription_payment( $amount = 0.0, $renewal_order, $re
 				throw new WC_Stripe_Exception( print_r( $response, true ), $localized_message );
 			}
 
-			do_action( 'wc_gateway_stripe_process_payment', $response, $renewal_order );
+			// Either the charge was successfully captured, or it requires further authentication.
+
+			if ( $is_authentication_required ) {
+				do_action( 'wc_gateway_stripe_process_payment_authentication_required', $renewal_order, $response );
 
+				$error_message = 'This transaction requires authentication.';
```
Comment (RadoslavGeorgiev) [inline includes/compat/class-wc-stripe-subs-compat.php]: This message should be internationalized.

**Comment by allendav** [comment]
@v18 - any update on this?

File: includes/compat/class-wc-stripe-subs-compat.php (hunk 1 above)
Comment (v18) [reply includes/compat/class-wc-stripe-subs-compat.php]: Fixed.

**Pull request merged**